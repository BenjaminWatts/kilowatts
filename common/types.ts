import { object, string, number, date, InferType, array, boolean } from "yup";

// Elexon Insights API

// reusable params

export type ElexonSettlementPeriodParams = {
  settlementDate: string;
  settlementPeriod: number;
};

export type ElexonRangeParams = {
  from: string; // a Date iso string in format YYYY-MM-DD
  to: string; // a Date iso string in format YYYY-MM-DD
  settlementPeriodFrom?: number;
  settlementPeriodTo?: number;
};

export const elexonInsightsPNDataRecordSchema = object({
  dataset: string().required().equals(["PN"]),
  settlementDate: string().required(),
  settlementPeriod: number().required(),
  timeFrom: string().required(),
  timeTo: string().required(),
  levelFrom: number().required(),
  levelTo: number().required(),
  nationalGridBmUnit: string().required(),
  bmUnit: string().nullable(),
});

export type ElexonInsightsPNDataRecord = InferType<
  typeof elexonInsightsPNDataRecordSchema
>;

type OptionalBmUnitParams = {
  bmUnits?: string[];
};

export type ElexonInsightsPnSpParams = ElexonSettlementPeriodParams &
  OptionalBmUnitParams;

export type ElexonInsightsPnRangeParams = ElexonRangeParams &
  OptionalBmUnitParams;

export type ElexonInsightsPnResponseRaw = {
  data: ElexonInsightsPNDataRecord[];
};

export const elexonInsightsPnResponseRawSchema = object({
  data: array(elexonInsightsPNDataRecordSchema).required(),
});

export type ElexonInsightsPnResponseRange = ElexonInsightsPNDataRecord[];

export type BmUnitId = string;
export type DateString = string;
export type LevelDict = Record<DateString, number>;
export type LevelPair = { time: string; level: number };
export type BmUnitValues = Record<BmUnitId, number>;
export type BmUnitLevelPairs = Record<BmUnitId, LevelPair[]>;

export type ElexonInsightsPnResponseParsed = BmUnitLevelPairs;

// export type ElexonInsightsAcceptancesDataRecord = {
//     settlementDate: string;
//     setttlementPeriodFrom: number;
//     setttlementPeriodTo: number;
//     timeFrom: string;
//     timeTo: string;
//     levelFrom: number;
//     levelTo: number;
//     nationalGridBmUnit: string;
//     bmUnit: string;
//     acceptanceNumber: number;
//     acceptanceTime: string;
//     deemedBoFlag: boolean;
//     soFlag: boolean;
//     storFlag: boolean;
//     rrFlag: boolean;
// }

const elexonInsightsAcceptancesDataRecordSchema = object({
  settlementDate: string().required(),
  settlementPeriodFrom: number().required(),
  settlementPeriodTo: number().required(),
  timeFrom: string().required(),
  timeTo: string().required(),
  levelFrom: number().required(),
  levelTo: number().required(),
  nationalGridBmUnit: string().required(),
  bmUnit: string().nullable(),
  acceptanceNumber: number().required(),
  acceptanceTime: string().required(),
  deemedBoFlag: boolean().required(),
  soFlag: boolean().required(),
  storFlag: boolean().required(),
  rrFlag: boolean().required(),
});
export type ElexonInsightsAcceptancesDataRecord = InferType<
  typeof elexonInsightsAcceptancesDataRecordSchema
>;

export type ElexonInsightsAcceptancesSpParams = ElexonInsightsPnSpParams;
export type ElexonInsightsAcceptancesRangeParams = ElexonRangeParams &
  OptionalBmUnitParams;

export const elexonInsightsAcceptancesResponseRawSchema = object({
  data: array(elexonInsightsAcceptancesDataRecordSchema).required(),
});

export type ElexonInsightsAcceptancesResponse = InferType<
  typeof elexonInsightsAcceptancesResponseRawSchema
>;

export type ElexonInsightsAcceptancesRangeResponse =
  ElexonInsightsAcceptancesDataRecord[];

export type ElexonInsightsAcceptancesParsedNoLevels = {
  bmUnit: string;
  acceptanceNumber: number;
  acceptanceTime: string;
  deemedBoFlag: boolean;
  soFlag: boolean;
  storFlag: boolean;
  rrFlag: boolean;
};

export type ElexonInsightsAcceptancesParsed =
  ElexonInsightsAcceptancesParsedNoLevels & {
    levels: LevelPair[];
  };

export type ElexonInsightsAcceptancesResponseParsed = Record<
  BmUnitId,
  ElexonInsightsAcceptancesParsed[]
>;

export type FuelType =
  | "gas"
  | "coal"
  | "nuclear"
  | "wind"
  | "hydro"
  | "biomass"
  | "solar"
  | "oil"
  | "interconnector"
  | "unknown"
  | "battery";

export type UnitGroupUnit = {
  bmUnit: string;
  name?: string;
};

export type UnitGroupDetails = {
  code?: string;
  name: string; // Pembroke
  coords?: {
    lat: number;
    lng: number;
  };
  fuelType: FuelType;
};

export type UnitGroup = {
  details: UnitGroupDetails;
  units: UnitGroupUnit[];
};

export type UnitGroupUnitLevel = {
  unit: UnitGroupUnit;
  level: number;
};

export type UnitGroupLevel = {
  details: UnitGroupDetails;
  units: UnitGroupUnitLevel[];
  level: number;
};

export type FuelTypeLevel = {
  name: FuelType;
  level: number;
  unitGroupLevels: UnitGroupLevel[];
};

export type UnitGroupsDict = Record<string, UnitGroup>;

export type BmUnitLevelValue = {
  id: string;
  level: number;
};

/// NG ESO API

export const ngEsoEmbeddedWindAndSolarForecastRawResponse = object({
  success: boolean().required().equals([true]),
  result: object({
    records: array(
      object({
        DATE_GMT: string().required(), // gives the date at midnight
        EMBEDDED_WIND_FORECAST: number().required(),
        EMBEDDED_WIND_CAPACITY: number().required(),
        TIME_GMT: string().required(),
        // _full_text: string().required(),
        SETTLEMENT_PERIOD: number().required(),
        SETTLEMENT_DATE: string().required(),
        EMBEDDED_SOLAR_FORECAST: number().required(),
        _id: number().required(),
        EMBEDDED_SOLAR_CAPACITY: number().required(),
      })
    ).required(),
  }).required(),
});

export type NgEsoEmbeddedWindAndSolarForecastRawResponse = InferType<
  typeof ngEsoEmbeddedWindAndSolarForecastRawResponse
>;

type EmbeddedForecastValue = {
    level: number;
    capacity: number;
}

export type NgEsoEmbeddedWindAndSolarForecastParsedResponse = {
    time: string;
    wind:EmbeddedForecastValue
    solar:EmbeddedForecastValue
}[]


// hooks
export type FuelTypeLiveCompleteness = {
  bm: boolean;
  embedded: boolean;
}

export type FuelTypeLiveHookResultLoading = {
  now: null
  isLoading: true;
  error: undefined;
  refetch: () => void;
  data: null;
  completeness: FuelTypeLiveCompleteness
}

export type FuelTypeLiveHookResultError = {
  now: null;
  isLoading: boolean;
  error: Error
  refetch: () => void;
  data: null;
  completeness: FuelTypeLiveCompleteness
}

export type FuelTypeLiveHookResultSuccess = {
  now: Date;
  isLoading: boolean;
  error: undefined;
  refetch: () => void;
  completeness: FuelTypeLiveCompleteness
  data: FuelTypeLevel[]
}

export type FuelTypeLiveHookResult = FuelTypeLiveHookResultSuccess | FuelTypeLiveHookResultError | FuelTypeLiveHookResultLoading;