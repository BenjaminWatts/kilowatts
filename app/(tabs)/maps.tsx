import React from 'react'
import log from '../../services/log';
import { Platform } from 'react-native';


export const MapTab =  () => {
    if (Platform.OS === 'web') {
        log.debug(`MapTab: web`)
        // return <NoMapsonWeb/>
        // Add web-specific code or handle the absence of the package here.
        const {MapViewWeb} = require('../../components/MapViewWeb')
        return <MapViewWeb/>
      } else {
        log.debug(`MapTab: mobile`)
        // Add mobile-specific code here, or else remove the import above.
        // You can also import other files to run at this point.
        // const {UnitGroupsMap} = require('../../components/MapView')
        // return <UnitGroupsMap/>
      }
    log.debug(`MapTab`)

    // return <UnitGroupsMap/>
}

export default MapTab