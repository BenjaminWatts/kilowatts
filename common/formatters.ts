
export const formatters = {
    mw: (x: number) => `${Math.round(x * 10)/10} MW`,
}
export default formatters