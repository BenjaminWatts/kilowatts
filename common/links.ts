
export const GITHUB_REPO_LINK = 'https://github.com/benjaminWatts/kilowatts'
export const GITHUB_UNIT_FILE_LINK = `https://github.com/BenjaminWatts/kilowatts/blob/main/assets/data/units.ts`