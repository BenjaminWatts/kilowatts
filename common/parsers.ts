import * as t from "./types";
import log from "../services/log";
import { get } from "react-native/Libraries/TurboModule/TurboModuleRegistry";

export const getBmUnits = (records: { bmUnit: string }[]): string[] => {
  let set = new Set<string>();
  records.forEach((r) => set.add(r.bmUnit));
  const output = Array.from(set);
  log.debug(
    `getBmUnits: ${output.length} units found from ${records.length} records`
  );
  return output;
};

export const levelDictToLevelPairs = (
  levelDict: t.LevelDict
): t.LevelPair[] => {
  let output: t.LevelPair[] = [];
  for (const time of Object.keys(levelDict)) {
    output.push({ time, level: levelDict[time] });
  }
  return output.sort((a, b) => a.time.localeCompare(b.time));
};

type IntervalRecord = {
  timeFrom: string;
  timeTo: string;
  levelFrom: number;
  levelTo: number;
};

export const intervalRecordToLevelDict = (r: IntervalRecord[]): t.LevelDict => {
  let levelDict: t.LevelDict = {};
  for (const x of r) {
    levelDict[x.timeFrom] = x.levelFrom;
    levelDict[x.timeTo] = x.levelTo;
  }
  return levelDict;
};

export const intervalRecordToLevelPairs = (
  r: IntervalRecord[]
): t.LevelPair[] => {
  const levelDict = intervalRecordToLevelDict(r);
  return levelDictToLevelPairs(levelDict);
};

export const interpolateLevelPair = (
  time: string,
  levelPairs: t.LevelPair[]
): number => {
  const exact = levelPairs.find((x) => x.time === time);
  if (exact) {
    log.debug(`interpolateLevelPair: exact match found for ${time}`);
  }
  const befores = levelPairs.filter((x) => x.time < time);
  const afters = levelPairs.filter((x) => x.time > time);
  if (befores.length === 0 || afters.length === 0) {
    log.debug(`interpolateLevelPair: no before or after found for ${time}`);
  }
  // interpolate on a linear basis
  const before = befores[befores.length - 1];
  const after = afters[0];
  const times = {
    before: new Date(before.time).getTime(),
    after: new Date(after.time).getTime(),
    output: new Date(time).getTime(),
  };
  const timeDiffs = {
    before: times.output - times.before,
    total: times.after - times.before,
    after: times.after - times.output,
  };
  const weights = {
    before: timeDiffs.before / timeDiffs.total,
    after: timeDiffs.after / timeDiffs.total,
  };
  const interp = before.level * weights.before + after.level * weights.after;
  return Math.round(interp);
};

type InterpolateBmUnitLevelPairsParams = {
  time: string;
  bmUnitLevelPairs: t.BmUnitLevelPairs;
  omitZero: boolean;
};

export const interpolateBmUnitLevelPairs = ({
  time,
  bmUnitLevelPairs,
  omitZero,
}: InterpolateBmUnitLevelPairsParams): t.BmUnitValues => {
  log.debug(`interpolateBmUnitLevelPairs: interpolating for ${time}`);
  let output: t.BmUnitValues = {};
  for (const bmUnit of Object.keys(bmUnitLevelPairs)) {
    const level = interpolateLevelPair(time, bmUnitLevelPairs[bmUnit]);
    if (level !== 0 || !omitZero) {
      output[bmUnit] = level;
    }
  }
  log.debug(
    `interpolateBmUnitLevelPairs: ${
      Object.keys(output).length
    } units found for ${time}`
  );
  return output;
};

export const sortDescendingBmUnitValues = (v: t.BmUnitValues) => {
  let bmUnits: { id: string; level: number }[] = [];
  for (const id of Object.keys(v)) {
    bmUnits.push({ id, level: v[id] });
  }
  return bmUnits.sort((a, b) => b.level - a.level);
};

export const getAcceptancesNoLevels = (
  x: t.ElexonInsightsAcceptancesDataRecord[]
): t.ElexonInsightsAcceptancesParsedNoLevels[] => {
  let output: Record<string, t.ElexonInsightsAcceptancesParsedNoLevels> = {};
  for (const a of x) {
    output[a.acceptanceNumber] = {
      bmUnit: a.bmUnit,
      acceptanceNumber: a.acceptanceNumber,
      acceptanceTime: a.acceptanceTime,
      deemedBoFlag: a.deemedBoFlag,
      soFlag: a.soFlag,
      storFlag: a.storFlag,
      rrFlag: a.rrFlag,
    };
  }
  return Object.values(output);
};

export const parseAcceptancesWithLevels = (
  x: t.ElexonInsightsAcceptancesDataRecord[]
): t.ElexonInsightsAcceptancesParsed[] =>
  getAcceptancesNoLevels(x).map((a) => {
    return {
      ...a,
      levels: intervalRecordToLevelPairs(
        x.filter((y) => y.acceptanceNumber === a.acceptanceNumber)
      ),
    };
  });


type CombinePnsAndAccsParams = {
  pns: t.ElexonInsightsPnResponseParsed;
  accs: t.ElexonInsightsAcceptancesResponseParsed;
}

export const combinePnsAndAccs = ({pns, accs}: CombinePnsAndAccsParams): t.BmUnitLevelPairs => {
  log.debug(`combinePnsAndAccs`)
  let output: t.BmUnitLevelPairs = {}
  console.log(accs)
  for (const bmUnit of Object.keys(pns)) {
    log.debug(`combinePnsAndAccs: combining ${bmUnit}`)
    if(!Object.keys(accs).includes(bmUnit)) {
      // log.debug(`combinePnsAndAccs: no acceptances found for ${bmUnit}`)
      output[bmUnit] = pns[bmUnit]
    } else {
      log.debug(`combinePnsAndAccs: acceptances found for ${bmUnit}`)
      let schedule: t.LevelPair[] = pns[bmUnit]
      for (const acc of accs[bmUnit]) {
        log.info(`combinePnsAndAccs: ${bmUnit} combining acceptance ${acc.acceptanceNumber}`)
        schedule = [
          ...schedule.filter(x => x.time < acc.levels[0].time),
          ...acc.levels,
          ...schedule.filter(x => x.time > acc.levels[acc.levels.length - 1].time)
        ]
      }
      log.debug(`updating schedule for ${bmUnit} from ${pns[bmUnit].length} to ${schedule.length} levels`)
      output[bmUnit] = schedule
    }
  }
  return output
}