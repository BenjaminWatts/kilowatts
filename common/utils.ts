import * as t from "./types";
import log from "../services/log";
import * as Updates from 'expo-updates';


export const londonTime = (date: Date) =>
  date.toLocaleString("en-GB", { timeZone: "Europe/London" });

export const londonTimeHHMMSS = (date: Date) =>
  londonTime(date).split(" ")[1];

export const getSettlementPeriod = (
  dateString?: string
): t.ElexonSettlementPeriodParams => {
  const date = londonTime(dateString ? new Date(dateString) : new Date());
  // in DD/MM/YYYY, HH:MM:SS format
  const [day, month, year, hour, minute] = date.split(/[\/\s,:]+/);
  const settlementDate = `${year}-${month}-${day}`;
  const settlementPeriod =
    parseInt(hour) * 2 + (parseInt(minute) >= 30 ? 1 : 0) + 1;
  log.debug(
    `getSettlementPeriod: ${settlementDate}:${settlementPeriod} for ${dateString}`
  );
  return {
    settlementDate,
    settlementPeriod,
  };
};

export const getSettlementDate = (dateString?: string): string => {
  const date = londonTime(dateString ? new Date(dateString) : new Date());
  // in DD/MM/YYYY, HH:MM:SS format
  const [day, month, year, hour, minute] = date.split(/[\/\s,:]+/);
  const settlementDate = `${year}-${month}-${day}`;
  return settlementDate;
};


/*
getCurrentYear
for rendering copyright noticed
  */
export const getCurrentYear = () => {
  return new Date().getFullYear();
};

/*
ALLOW_LINK_FUELTYPES
These are fuel types for which it is allowed to filter by
*/
export const ALLOW_LINK_FUELTYPES: t.FuelType[] = [
  "wind",
  "hydro",
  "nuclear",
  "biomass",
  "gas",
  "coal",
  "interconnector"
];


/*
getVersion
using expo updates library to get the native and js version
*/
export const getVersion = () => {
  return {
    createdAt: Updates.createdAt
  }
}