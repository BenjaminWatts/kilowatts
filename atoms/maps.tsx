import React from "react";
import MV, { MapViewProps, Marker, PROVIDER_GOOGLE } from "react-native-maps";
import {
  UnitGroupMapProps,
  UnitGroupMarker,
  UnitsGroupMapProps,
} from "../common/types";
import { Platform, StyleSheet } from "react-native";
import formatters from "../common/formatters";
import { FuelTypeIcon } from "./icons";
import { getUnitGroupUrl } from "../services/nav";
import { useRouter } from "expo-router";

const getMapProvider = () => {
  if (Platform.OS === "android") {
    return PROVIDER_GOOGLE;
  }
  return undefined;
};

const MapView = (props: MapViewProps) => (
  <MV
    provider={getMapProvider()}
    {...props}
    // cacheEnabled={true}
    liteMode={true}
    mapType={Platform.select({
      ios: "mutedStandard",
      android: "satellite",
    })}
  />
);

export const UnitGroupMap: React.FC<UnitGroupMapProps> = ({ ug }) => {
  console.log(`UnitGroupMap: ${ug.details.name}`);
  const { coords, name, fuelType } = ug.details;
  if (!coords || !name) return <></>;
  const coordinate = {
    latitude: coords.lat,
    longitude: coords.lng,
  };
  return (
    <MapView
      scrollEnabled={false}
      style={styles.mapCardContainer}
      initialRegion={{
        ...coordinate,
        latitudeDelta: 0.0922 * 15,
        longitudeDelta: 0.0421 * 15,
      }}
    >
      <Marker
        coordinate={coordinate}
        title={ug.details.name}
        description={ug.details.fuelType}
      >
        <FuelTypeIcon fuelType={fuelType} size={20} />
      </Marker>
    </MapView>
  );
};

export const UnitsGroupMap: React.FC<UnitsGroupMapProps> = ({ markers }) => {
  const router = useRouter();

  return (
    <MapView
      scrollEnabled={false}
      zoomEnabled={false}
      rotateEnabled={false}
      zoomTapEnabled={false}
      style={styles.mapCardContainer}
      initialRegion={{
        latitude: 54.5,
        longitude: -2,
        latitudeDelta: 9,
        longitudeDelta: 12,
      }}
    >
      {markers.map((marker) => (
        <Marker
          key={marker.code}
          onPress={() => {
            const url = getUnitGroupUrl(marker);
            if (!url) return;
            router.push(url as any);
          }}
          {...marker}
          description={`${formatters.fuelType(marker.fuelType)}`}
        >
          <FuelTypeIcon fuelType={marker.fuelType} size={20} />
        </Marker>
      ))}
    </MapView>
  );
};

const styles = StyleSheet.create({
  mapCardContainer: {
    display: "flex",
    flex: 1,
    height: "50%",
  },
});
