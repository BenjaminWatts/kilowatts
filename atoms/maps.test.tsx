import { UnitGroup } from "../common/types";
import * as m from "./maps";
import { render, screen } from "@testing-library/react-native";

//mock expo-router
jest.mock("expo-router", () => ({
  useRouter: () => jest.fn(),
}));

const mockUg: UnitGroup = {
  details: {
    name: "My Powerstation",
    coords: {
      lat: 50,
      lng: -3,
    },
    fuelType: "coal",
  },
  units: [],
};

describe("atoms/maps/UnitGroupMap", () => {
  test("can render", () => {
    render(<m.UnitGroupMap ug={mockUg} />);
    console.log = jest.fn();
    expect(screen.debug({})).toMatchSnapshot();
  });
});
