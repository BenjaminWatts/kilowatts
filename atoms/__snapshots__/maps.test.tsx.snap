// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/maps/UnitGroupMap can render 1`] = `undefined`;
